blog: Serving Models with MLEM (#3623)
iterative/dvc.org: content/blog/2022-06-28-serving-models-with-mlem.md

---

> title: Serving Models with MLEM

Reviewer: ```suggestion
title: Serving ML Models with MLEM
```

date: 2022-06-28
description: >

>   Once you have a model that's ready for production, getting it out can be a
>   complicated step. In this tutorial, we're going to use MLEM to deploy a model

Reviewer: ```suggestion
  Once you have a machina learning model that's ready for production, getting it out can be
  complicated. In this tutorial, we're going to use MLEM to deploy a model
```

  to an API.
descriptionLong: >

>   Getting models out to our end users is usually a difficult task. You have to
>   make sure it's in the correct format for the environment and for the way it
>   will be consumed. We're going to walk through how you can use MLEM to save
>   your model and serve it through an API.

Reviewer: ```suggestion
  Getting a machine learning model out to our end users can be a difficult task. You have to
  make sure it's in the correct format for the environment and for the way it
  will be consumed. We're going to walk through how you can use MLEM to save
  your model and serve it with a web API.
```

> can be a difficult task

Or if we could say "doesn't have to be a difficult task", that could be a better hook.

picture: 2022-06-28/mlem-deploys.png
pictureComment: Serving Models with MLEM
author: milecia_mcgregor
commentsUrl: https://discuss.dvc.org/t/syncing-data-to-aws-s3/1192
tags:
  - MLOps
  - DVC
  - Git
  - MLEM
  - Collaboration
---

Training a machine learning model is only one step in the process of getting

> something useful out to the end users. The goal of machine learning projects to

Reviewer 2: ...goal of machine learning projects *is* to ....

> Training a machine learning model is only one step in the process of getting
> something useful out to the end users. The goal of machine learning projects to
> make an accurate model that provides predictions based on the data received from
> a production environment. That could mean making predictions on data coming from
> an API or do some batch predictions.

Reviewer: ```suggestion
Training an ML model is only one step in the process of getting
something useful out to the end users. The goal of machine learning projects is to
make an accurate model that provides predictions based on the data received from
a production environment. That could mean data coming from
an API or batch processing, for example.
```

This is a bit low-level though. And at the same time generic. "make an accurate model", "production environment", "data coming from an API": can we bring it down to earth? Or is it for a technical audience

Reviewer: p.s. I don't think all ML models make predictions though. Could be all sorts of tasks?

> Either way, you'll need to save your trained and validated model in format
> that's consumable by other systems. That's why we'll be covering how to serve
> models through an endpoint with [MLEM](https://mlem.ai/). We'll start by
> training a simple model. You can get the repo we're working with
> [here](https://github.com/iterative/stale-model-example/tree/mlem-serve).

Reviewer: ```suggestion
Either way, you'll need to save your trained and validated model in format
that's consumable by other systems. That's why we'll be covering how to serve
models through a [REST](https://restfulapi.net/) endpoint with [MLEM](https://mlem.ai/).

> You can get the repo we're working with
> [here](https://github.com/iterative/stale-model-example/tree/mlem-serve).
```

## Train a model

There are instructions in the project
[README](https://github.com/iterative/stale-model-example/tree/mlem-serve#readme)

> on how to get everything you need installed and running. Make sure you run a few

Reviewer: I think we should explain that it's a DVC project and that MLEM and DVC play well together.

 ───────────────────────────────────────────────────────────────────────────────────────────
  workspace                  0.72417   0.38926   20210428     250           75
  mlem-serve                 0.76681   0.38867   20210428     300           75
  ├── c8d952b [exp-4b026]    0.72417   0.38926   20210428     250           75
  ├── da78e7c [exp-7a2e2]    0.74454    0.5528   20210428     450           75
  ├── b39470f [exp-5ebf6]    0.42526   0.82722   20210428     400           28
  ├── 03c6d57 [exp-f8c3a]    0.74454    0.5528   20210428     400           75
  ├── 75b1b3c [exp-2ba75]    0.67913   0.38207   20210428     350           75
  └── 40b0075 [exp-e0dc1]    0.76681   0.38867   20210428     300           75
 ───────────────────────────────────────────────────────────────────────────────────────────
```

Of course you would run many more experiments in order to determine the best
model for production, but in this example, the metrics for `exp-5ebf6` look
pretty good. So we'll take this model and get it ready to serve to production.

## Save the model

First, we need to install the `MLEM` package with:

```dvc

> $ pip install 'mlem[fastapi]'

Reviewer: Could we do this first though, install MLEM, plug it in the existing code, and THEN `dvc exp run` to train and select a model. (Idea being: Start with MLEM since it's a MLEM post.)

Author: I made a pretty big update to the whole structure and moved the DVC stuff into an info box. I think that should clear the way for MLEM to stay the focus.

```

Then, inside of the `train.py` script, we need to add the MLEM import to save
the models as we experiment.

```python
# train.py

import os
import pickle5 as pickle
import sys
import yaml

from mlem.api import save
import numpy as np
from sklearn.ensemble import RandomForestClassifier
...
```

Then you can add the `save` function to the end of the training script.

```python
# train.py
...
clf = RandomForestClassifier(
    n_estimators=n_est, min_samples_split=min_split, n_jobs=2, random_state=seed
)

clf.fit(x, labels)

save(
    clf,
    "clf",
    sample_data=x,
    description="Random Forest Classifier",
)
```

_Note:_ If you want to see what is happening in the DVC pipeline, check out the

> `dvc.yaml` file in the project.

Reviewer: Shat do you mean here? Some step seems to be missing. Are you assuming `exp run` is used after the edit? It's not indicated. (This may be irrelevant if we change the section order as I suggested above.)

There will be two new files in your repo: `clf` and `clf.mlem`. You'll be able
to load the model file into the evaluation script and make sure it gives you the

> performance you need.

Reviewer: Put clf.mlem in Git? To sell GitOps advantage

Author: Made that update to the repo!

## Load the model for evaluation

We're going to load the MLEM model into our `src/evaluate.py` script. To do
that, you'll need to add the following import.

```python
# evaluate.py

import json
import math
import os
import sys

import pickle5 as pickle
import sklearn.metrics as metrics
import numpy as np
from mlem.api import apply
...
```

Now we can use the [`apply` function](https://mlem.ai/doc/api-reference/apply)
to make predictions with the model.

```python
# evaluate.py
...
x = matrix.iloc[:,1:11].values

cleaned_x = np.where(np.isnan(x), 0, x)
labels_pred = apply(model_file, cleaned_x, method="predict")

predictions_by_class = apply(model_file, cleaned_x, method="predict_proba")
predictions = predictions_by_class[:, 1]
...
```

The `predict` and `predict_proba` are methods available from the model and they
are used to get new predicted values and their probabilities for evaluation.

> This, alond with everything else in the script, is how we get those metrics are

Reviewer 2: alond -> along

shown in the experiments table. Once you've validated the model and decided that
it performs well, it's ready to go to production.

## Serve the model to production

There are a couple of ways you can do this with MLEM:

- Create a Python package

> - Serve the model with [FastAPI](https://fastapi.tiangolo.com/)

Reviewer: 1. Takes a while to get here. Maybe we should skip the whole training?
2. The title only mentioned the web API but we're also introducing the Python package here. OK to keep but after `mlem serve` I think, and include in title or at least post descriptions, I think.

Author: 1. Yeah, I made the training/evaluate steps an info box that can be skipped.
2. I moved the Python package to make it the second way to serve a model.

```

Now you can import this to your Flask API like so.

```python
# api.py
import os
from flask import Flask, jsonify, request
from flask_sqlalchemy import SQLAlchemy
from flask_migrate import Migrate
from flask_cors import CORS
from dotenv import load_dotenv
import bike_predictor
...
```

You can then use the `predict` method on new data and run any other tasks you
need to in the API.

```python
# api.py
...
new_event = EventsModel(
    title=data["title"],
    date=data["date"],
    time=data["time"],
    location=data["location"],
    description=data["description"],
)

db.session.add(new_event)
db.session.commit()

bike_predictor.predict(new_event)
...
```

Then you can test this API out locally by running.

```dvc
$ python src/api.py
```

This will start up a local server on port 5000 and you'll be able to see your
model in action. From here, this can be deployed to any cloud environment as
long as you remember to include and install the model package.

### Serve with FastAPI

If you don't have an API to work with and don't need a Python package, you can
serve your model quickly using FastAPI with this command.

```dvc

> $ mlem serve clf fastapi

Reviewer: The main part of the blog post is almost at the very bottom :B
